File: demo/src/main/java/com/example/Portfolio.java
package com.example;


import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.io.FileWriter;
import java.io.IOException;
import java.io.BufferedReader;
import java.io.FileReader;
import java.time.format.DateTimeFormatter;

import com.example.AssetBuilder;
import com.example.FinancialAsset;
import com.example.IntermediaryBuilder;
import com.example.SnapshotBuilder;
import com.example.Stock;


public class Portfolio {
    private List<FinancialAsset> assets;
    private List<FinancialIntermediary> intermediaries;
    private List<Snapshot> snapshots = new ArrayList<>();

    public Portfolio() {
        assets = new ArrayList<>();
        intermediaries = new ArrayList<>();
    }

    // CREATE: Add a financial asset
    public void addAsset(FinancialAsset asset) {
        assets.add(asset);
    }

    // READ: Get an asset by name
    public FinancialAsset getAsset(String name) {
        for (FinancialAsset asset : assets) {
            if (asset.getName().equals(name)) {
                return asset;
            }
        }
        return null;  // Asset not found
    }

    public List<FinancialAsset> getAssets() {
        return assets;
    }

    // UPDATE: Update asset value by name
    public boolean updateAsset(String name, float newValue) {
        FinancialAsset asset = getAsset(name);
        if (asset != null) {
            switch(asset.getClass().getSimpleName()){
                case "Stock":
                    FinancialAsset stock = new AssetBuilder()
                            .setName(asset.getName())
                            .setValue(newValue)
                            .setTickerSymbol(((Stock) asset).getTickerSymbol())
                            .setShares(((Stock) asset).getShares())
                            .setDividendYield(((Stock) asset).getDividendYield())
                            .setType("Stock")
                            .build();
                    assets.remove(asset);
                    assets.add(stock);
                    return true;
                case "Bond":
                    FinancialAsset bond = new AssetBuilder()
                            .setName(asset.getName())
                            .setValue(newValue)
                            .setInterestRate(((Bond) asset).getInterestRate())
                            .setMaturityDate(((Bond) asset).getMaturityDate())
                            .setType("Bond")
                            .build();
                    assets.remove(asset);
                    assets.add(bond);
                    return true;
                case "MutualFund":
                    FinancialAsset mutualFund = new AssetBuilder()
                            .setName(asset.getName())
                            .setValue(newValue)
                            .setExpenseRatio(((MutualFund) asset).getExpenseRatio())
                            .setType("MutualFund")
                            .build();
                    assets.remove(asset);
                    assets.add(mutualFund);
                    return true;
            }
        }
        return false;  // Asset not found
    }

    // DELETE: Remove an asset by name
    public boolean removeAsset(String name) {
        FinancialAsset asset = getAsset(name);
        if (asset != null) {
            assets.remove(asset);
            return true;
        }
        return false;  // Asset not found
    }

    // CREATE: Add an intermediary
    public void addIntermediary(FinancialIntermediary intermediary) {
        intermediaries.add(intermediary);
    }

    // READ: Get an intermediary by name
    public FinancialIntermediary getIntermediary(String name) {
        for (FinancialIntermediary intermediary : intermediaries) {
            if (intermediary.getIntermediaryName().equals(name)) {
                return intermediary;
            }
        }
        return null;  // Intermediary not found
    }

    public List<FinancialIntermediary> getIntermediaries() {
        return intermediaries;
    }

    // UPDATE: Update intermediary name by old name
    public boolean updateIntermediary(String oldName, String newName) {
        FinancialIntermediary intermediary = getIntermediary(oldName);
        if (intermediary != null) {
            switch(intermediary.getClass().getSimpleName()){
                case "Broker":
                    FinancialIntermediary broker = new IntermediaryBuilder()
                            .setName(newName)
                            .setType("Broker")
                            .build();
                    intermediaries.remove(intermediary);
                    intermediaries.add(broker);
                    return true;
                case "Bank":
                    FinancialIntermediary bank = new IntermediaryBuilder()
                            .setName(newName)
                            .setType("Bank")
                            .build();
                    intermediaries.remove(intermediary);
                    intermediaries.add(bank);
                    return true;
                case "MutualFundManager":
                    FinancialIntermediary mutualFundManager = new IntermediaryBuilder()
                            .setName(newName)
                            .setType("MutualFundManager")
                            .build();
                    intermediaries.remove(intermediary);
                    intermediaries.add(mutualFundManager);
                    return true;
            }
        }
        return false;  // Intermediary not found
    }

    // DELETE: Remove an intermediary by name
    public boolean removeIntermediary(String name) {
        FinancialIntermediary intermediary = getIntermediary(name);
        if (intermediary != null) {
            intermediaries.remove(intermediary);
            return true;
        }
        return false;  // Intermediary not found
    }

    public void displayPortfolio() {
        System.out.println("Portfolio Assets:");
        for (FinancialAsset asset : assets) {
            asset.displayAsset();
        }

        System.out.println("\nIntermediaries:");
        for (FinancialIntermediary intermediary : intermediaries) {
            intermediary.displayIntermediary();
        }
    }

    public float calculateTotalAnnualReturn() {
        float totalReturn = 0;
        for (FinancialAsset asset : assets) {
            totalReturn += asset.calculateAnnualReturn();
        }
        return totalReturn;
    }

    public void takeSnapshot() {
        for (FinancialAsset asset : assets) {
            LocalDateTime timestamp = LocalDateTime.now();
            String timeString = timestamp.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
            Snapshot snapshot = new SnapshotBuilder()
                    .setAssetName(asset.getName())
                    .setAssetValue(asset.getValue())
                    .setTimestamp(timeString)
                    .build();
            snapshots.add(snapshot);
        }
        System.out.println("Snapshot taken.");
    }

    public void addSnapshot(Snapshot snapshot) {
        snapshots.add(snapshot);
    }

    public void displaySnapshots() {
        for (Snapshot snapshot : snapshots) {
            System.out.println("Snapshot - Asset: " + snapshot.getAssetName() +
                    ", Value: $" + snapshot.getAssetValue() +
                    ", Time: " + snapshot.getTimestamp());
        }
    }

    public List<Snapshot> getSnapshots() {
        return snapshots;
    }

    public void displaySnapshotsForAsset(String assetName, LocalDateTime startDate, LocalDateTime endDate) {
        System.out.println("Snapshots for asset: " + assetName);
        for (Snapshot snapshot : snapshots) {
            if (snapshot.getAssetName().equals(assetName)) {
                LocalDateTime timestamp = LocalDateTime.parse(snapshot.getTimestamp(), DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
                if (timestamp.isAfter(startDate) && timestamp.isBefore(endDate)) {
                    snapshot.displaySnapshot();
                }
            }
        }
    }

    public void saveState(String filename) throws IOException {
        try (FileWriter writer = new FileWriter(filename)) {
            // Save financial assets
            for (FinancialAsset asset : assets) {
                //writer.write(asset.getClass().getSimpleName() + "," + asset.getName() + "," + asset.getValue() + "\n");
                switch(asset.getClass().getSimpleName()){
                    case "Stock":
                        writer.write("Stock," + asset.getName() + "," + asset.getValue() + "," + ((Stock) asset).getTickerSymbol() + "," + ((Stock) asset).getShares() + "," + ((Stock) asset).getDividendYield() + "\n");
                        break;
                    case "Bond":
                        writer.write("Bond," + asset.getName() + "," + asset.getValue() + "," + ((Bond) asset).getInterestRate() + "," + ((Bond) asset).getMaturityDate() + "\n");
                        break;
                    case "MutualFund":
                        writer.write("MutualFund," + asset.getName() + "," + asset.getValue() + "," + ((MutualFund) asset).getExpenseRatio() + "\n");
                        break;
                }
            }

            // Save snapshots
            for (Snapshot snapshot : snapshots) {
                writer.write("Snapshot," + snapshot.getAssetName() + "," + snapshot.getAssetValue() + "," + snapshot.getTimestamp() + "\n");
            }

            // Save intermediaries
            for (FinancialIntermediary intermediary : intermediaries) {
                switch (intermediary.getClass().getSimpleName()) {
                    case "Broker":
                        writer.write("Broker," + intermediary.getIntermediaryName() + "\n");
                        break;
                    case "Bank":
                        writer.write("Bank," + intermediary.getIntermediaryName() + "\n");
                        break;
                    case "MutualFundManager":
                        writer.write("MutualFundManager," + intermediary.getIntermediaryName() + "\n");
                        break;
                }
            }

            System.out.println("Portfolio saved to " + filename);
        }
    }

    public void loadState(String filename) throws IOException {

        assets.clear();
        intermediaries.clear();
        snapshots.clear();
        
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String[] data = line.split(",");

                switch (data[0]) {
                    case "Stock":
                        //assets.add(new Stock(data[1], Float.parseFloat(data[2]), data[3], Integer.parseInt(data[4]), Float.parseFloat(data[5])));
                        assets.add(new AssetBuilder()
                                .setName(data[1])
                                .setValue(Float.parseFloat(data[2]))
                                .setTickerSymbol(data[3])
                                .setShares(Integer.parseInt(data[4]))
                                .setDividendYield(Float.parseFloat(data[5]))
                                .setType("Stock")
                                .build());
                        break;
                    case "Bond":
                        //assets.add(new Bond(data[1], Float.parseFloat(data[2]), 0.03f, "2030-01-01"));
                        assets.add(new AssetBuilder()
                                .setName(data[1])
                                .setValue(Float.parseFloat(data[2]))
                                .setInterestRate(Float.parseFloat(data[3]))
                                .setMaturityDate(data[4])
                                .setType("Bond")
                                .build());
                        break;
                    case "MutualFund":
                        //assets.add(new MutualFund(data[1], Float.parseFloat(data[2]), 0.02f));
                        assets.add(new AssetBuilder()
                                .setName(data[1])
                                .setValue(Float.parseFloat(data[2]))
                                .setExpenseRatio(Float.parseFloat(data[3]))
                                .setType("MutualFund")
                                .build());
                        break;
                    case "Snapshot":
                        //snapshots.add(new Snapshot(data[1], Float.parseFloat(data[2]), LocalDateTime.parse(data[3])));
                        snapshots.add(new SnapshotBuilder()
                                .setAssetName(data[1])
                                .setAssetValue(Float.parseFloat(data[2]))
                                .setTimestamp(data[3])
                                .build());
                        break;
                    case "Broker":
                        intermediaries.add(new IntermediaryBuilder()
                                .setName(data[1])
                                .setType("Broker")
                                .build());
                        break;
                    case "Bank":
                        intermediaries.add(new IntermediaryBuilder()
                                .setName(data[1])
                                .setType("Bank")
                                .build());
                        break;
                    case "MutualFundManager":
                        intermediaries.add(new IntermediaryBuilder()
                                .setName(data[1])
                                .setType("MutualFundManager")
                                .build());
                        break;
                }
            }
            System.out.println("Portfolio loaded from " + filename);
        }
    }
}

File: src/Portfolio.cpp
#include "Portfolio.h"
#include "Snapshot.h" 
#include "Stock.h"
#include "Bond.h"        
#include "MutualFund.h"  
#include "Broker.h"  
#include "Bank.h"
#include "MutualFundManager.h"    
#include <ctime>
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>

Portfolio::~Portfolio() {
    for (auto asset : assets) delete asset;
    for (auto intermediary : intermediaries) delete intermediary;
}

// add asset to portfolio
void Portfolio::addAsset(FinancialAsset* asset) {
    assets.push_back(asset);
}

// get asset 
FinancialAsset* Portfolio::getAsset(const std::string& name) const {
    for (auto asset : assets) {
        if (asset->getName() == name) {
            return asset;
        }
    }
    return nullptr;
}

// update asset 
bool Portfolio::updateAsset(const std::string& name, float newValue) {
    FinancialAsset* asset = getAsset(name);
    if (asset) {
        asset->setValue(newValue);
        return true;
    }
    return false;
}

// sell an asset 
bool Portfolio::sellAsset(const std::string& name) {
    for (auto it = assets.begin(); it != assets.end(); ++it) {
        if ((*it)->getName() == name) {
            delete *it;  // freeing asset memory
            assets.erase(it);  // removing from the vector
            return true;
        }
    }
    return false;
}

// add intermediary to the portfolio
void Portfolio::addIntermediary(FinancialIntermediary* intermediary) {
    intermediaries.push_back(intermediary);
}

// get intermediary 
FinancialIntermediary* Portfolio::getIntermediary(const std::string& name) const {
    for (auto intermediary : intermediaries) {
        if (intermediary->getIntermediaryName() == name) {
            return intermediary;
        }
    }
    return nullptr;
}

// update intermediary 
bool Portfolio::updateIntermediary(const std::string& name, const std::string& newIntermediaryName) {
    FinancialIntermediary* intermediary = getIntermediary(name);
    if (intermediary) {
        intermediary->setIntermediaryName(newIntermediaryName);
        return true;
    }
    return false;
}

// remove intermediary 
bool Portfolio::removeIntermediary(const std::string& name) {
    for (auto it = intermediaries.begin(); it != intermediaries.end(); ++it) {
        if ((*it)->getIntermediaryName() == name) {
            delete *it; 
            intermediaries.erase(it);  
            return true;
        }
    }
    return false;  
}

// take snapshot of all current assets
void Portfolio::takeSnapshot() {
    std::time_t currentTime = std::time(nullptr);
    for (const auto& asset : assets) {
        Snapshot snapshot(asset->getName(), asset->getValue(), currentTime);
        assetSnapshots.push_back(snapshot);
    }
}

// display all stored snapshots
void Portfolio::displaySnapshots() const {
    for (const auto& snapshot : assetSnapshots) {
        snapshot.displaySnapshot();
    }
}

// snapshots of a specific asset between start and end dates
void Portfolio::displaySnapshotsByDate(const std::string& assetName, std::time_t startDate, std::time_t endDate) const {
    bool found = false;
    for (const auto& snapshot : assetSnapshots) {
        if (snapshot.getAssetName() == assetName &&
            snapshot.getSnapshotDate() >= startDate &&
            snapshot.getSnapshotDate() <= endDate) {
            
            snapshot.displaySnapshot();
            found = true;
        }
    }

    if (!found) {
        std::cout << "No snapshots found for asset " << assetName 
                  << " in the given date range." << std::endl;
    }
}

float Portfolio::calculateTotalAnnualReturn() const {
    float totalReturn = 0;
    for (const auto& asset : assets) {
        totalReturn += asset->calculateAnnualReturn();
    }
    return totalReturn;
}

void Portfolio::displayPortfolio() const {
    for (const auto& asset : assets) {
        asset->displayAsset();
    }

    for (const auto& intermediary : intermediaries) {
        intermediary->displayIntermediary();
    }
}

void Portfolio::saveState(const std::string& filename) const {
    std::ofstream outFile(filename);

    if (!outFile) {
        std::cerr << "Error opening file for writing: " << filename << std::endl;
        return;
    }

    // saving assets with specific attributes
    for (const auto& asset : assets) {
        if (auto stock = dynamic_cast<Stock*>(asset)) {
            outFile << "Stock," << stock->getName() << "," << stock->getValue() << ","
                    << stock->getShares() << "," << stock->getDividendYield() << "\n";
        } else if (auto bond = dynamic_cast<Bond*>(asset)) {
            outFile << "Bond," << bond->getName() << "," << bond->getValue() << ","
                    << bond->getInterestRate() << "," << bond->getMaturityDate() << "\n";
        } else if (auto mutualFund = dynamic_cast<MutualFund*>(asset)) {
            outFile << "MutualFund," << mutualFund->getName() << "," << mutualFund->getValue() << ","
                    << mutualFund->getExpenseRatio() << "\n";
        }
    }

    // delimiter to separate assets and snapshots
    outFile << "--- Snapshots ---\n";

    // snapshots
    for (const auto& snapshot : assetSnapshots) {
        outFile << snapshot.getAssetName() << "," << snapshot.getAssetValue() << ","
                << snapshot.getSnapshotDate() << "\n";
    }

    // intermediaries
    outFile << "--- Intermediaries ---\n";
    for (const auto& intermediary : intermediaries) {
        if (auto broker = dynamic_cast<Broker*>(intermediary)) {
            outFile << "Broker," << broker->getIntermediaryName() << "\n";
        } else if (auto bank = dynamic_cast<Bank*>(intermediary)) {
            outFile << "Bank," << bank->getIntermediaryName() << "\n";
        } else if (auto mutualFundManager = dynamic_cast<MutualFundManager*>(intermediary)) {
            outFile << "MutualFundManager," << mutualFundManager->getIntermediaryName() << "\n";
        }
    }

    outFile.close();
    std::cout << "Portfolio state saved to " << filename << "\n";
}


std::vector<std::string> split(const std::string& str, char delimiter) {
    std::vector<std::string> tokens;
    std::string token;
    std::istringstream tokenStream(str);
    while (std::getline(tokenStream, token, delimiter)) {
        tokens.push_back(token);
    }
    return tokens;
}

void Portfolio::loadState(const std::string& filename) {
    std::ifstream inFile(filename);

    if (!inFile) {
        std::cerr << "Error opening file: " << filename << std::endl;
        return;
    }

    std::string line;
    std::vector<std::string> tokens;

    // clear portfolio data
    assets.clear();
    assetSnapshots.clear();
    intermediaries.clear();

    // assets
    std::cout << "Loading assets..." << std::endl;
    while (std::getline(inFile, line)) {
        if (line == "--- Snapshots ---") {
            break;  // end assets
        }

        tokens = split(line, ',');
        if (tokens.size() > 1) {
            std::string assetType = tokens[0];
            std::string name = tokens[1];
            float value = std::stof(tokens[2]);

            if (assetType == "Stock") {
                int shares = std::stoi(tokens[3]);
                float dividendYield = std::stof(tokens[4]);
                assets.push_back(new Stock(name, value, "", shares, dividendYield));
                std::cout << "Loaded Stock: " << name << ", value: " << value << "\n";
            } else if (assetType == "Bond") {
                float interestRate = std::stof(tokens[3]);
                std::string maturityDate = tokens[4];
                assets.push_back(new Bond(name, value, interestRate, maturityDate));
                std::cout << "Loaded Bond: " << name << ", value: " << value << "\n";
            } else if (assetType == "MutualFund") {
                float expenseRatio = std::stof(tokens[3]);
                assets.push_back(new MutualFund(name, value, expenseRatio));
                std::cout << "Loaded MutualFund: " << name << ", value: " << value << "\n";
            }
        }
    }

    // snapshots
    std::cout << "Loading snapshots..." << std::endl;
    while (std::getline(inFile, line)) {
        if (line == "--- Intermediaries ---") {
            break;  // end snapshots
        }

        tokens = split(line, ',');
        if (tokens.size() == 3) {
            std::string name = tokens[0];
            float value = std::stof(tokens[1]);
            std::time_t snapshotDate = std::stol(tokens[2]);

            assetSnapshots.push_back(Snapshot(name, value, snapshotDate));
            std::cout << "Loaded Snapshot: " << name << ", value: " << value << "\n";
        }
    }

    // intermediaries
    std::cout << "Loading intermediaries..." << std::endl;
    while (std::getline(inFile, line)) {
        tokens = split(line, ',');
        if (tokens.size() == 2) {
            std::string intermediaryType = tokens[0];
            std::string name = tokens[1];

            if (intermediaryType == "Broker") {
                intermediaries.push_back(new Broker(name));
            } else if (intermediaryType == "Bank") {
                intermediaries.push_back(new Bank(name));
            } else if (intermediaryType == "MutualFundManager") {
                intermediaries.push_back(new MutualFundManager(name));
            }

            std::cout << "Loaded Intermediary: " << intermediaryType << " - " << name << "\n";
        }
    }

    inFile.close();
    std::cout << "Portfolio state loaded from " << filename << "\n";
}
